test/ssl_handshake_rtt_test.c

```diff
@@ -0,0 +1,138 @@
+
+#include "helpers/ssltestlib.h"
+#include "testutil.h"
+#include "testutil/output.h"
+#include "internal/ktls.h"
+#include "../ssl/ssl_local.h"
+#include "../ssl/statem/statem_local.h"
+
+static OSSL_LIB_CTX *libctx = NULL;
+static char *cert = NULL;
+static char *privkey = NULL;
+
+/*
+ * Test 0: Clientside handshake RTT (TLSv1.2)
+ * Test 1: Serverside handshake RTT (TLSv1.2)
+ * Test 2: Clientside handshake RTT (TLSv1.3)
+ * Test 3: Serverside handshake RTT (TLSv1.3)
+ * Test 4: Clientside handshake RTT with Early Data (TLSv1.3)
+ */
+static int test_handshake_rtt(int tst)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    SSL_CONNECTION *s = NULL;
+    OSSL_STATEM *st = NULL;
+    uint64_t rtt;
+
+#ifdef OPENSSL_NO_TLS1_2
+    if (tst <= 1)
+        return 1;
+#endif
+#ifdef OSSL_NO_USABLE_TLS1_3
+    if (tst >= 2)
+        return 1;
+#endif
+
+    if (!TEST_true(create_ssl_ctx_pair(libctx, TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_VERSION,
+                                       (tst <= 1) ? TLS1_2_VERSION
+                                                  : TLS1_3_VERSION,
+                                       &sctx, &cctx, cert, privkey))
+            || !TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
+                                             NULL, NULL)))
+        goto end;
+
+    s = SSL_CONNECTION_FROM_SSL(tst % 2 == 0 ? clientssl : serverssl);
+    if (s == NULL || (st = &s->statem) == NULL)
+        return 0;
+
+    /* implicitly set handshake rtt with a delay */
+    switch (tst) {
+    case 0:
+        st->hand_state = TLS_ST_CW_CLNT_HELLO;
+        ossl_statem_client_write_transition(s);
+        OSSL_sleep(1);
+        st->hand_state = TLS_ST_CR_SRVR_DONE;
+        ossl_statem_client_write_transition(s);
+        break;
+    case 1:
+        st->hand_state = TLS_ST_SW_SRVR_DONE;
+        ossl_statem_server_write_transition(s);
+        OSSL_sleep(1);
+        st->hand_state = TLS_ST_SR_FINISHED;
+        ossl_statem_server_write_transition(s);
+        break;
+    case 2:
+        st->hand_state = TLS_ST_CW_CLNT_HELLO;
+        ossl_statem_client_write_transition(s);
+        OSSL_sleep(1);
+        st->hand_state = TLS_ST_CR_SRVR_DONE;
+        ossl_statem_client_write_transition(s);
+        break;
+    case 3:
+        st->hand_state = TLS_ST_SW_SRVR_DONE;
+        ossl_statem_server_write_transition(s);
+        OSSL_sleep(1);
+        st->hand_state = TLS_ST_SR_FINISHED;
+        ossl_statem_server_write_transition(s);
+        break;
+    case 4:
+        st->hand_state = TLS_ST_EARLY_DATA;
+        ossl_statem_client_write_transition(s);
+        OSSL_sleep(1);
+        st->hand_state = TLS_ST_CR_SRVR_DONE;
+        ossl_statem_client_write_transition(s);
+        break;
+    }
+
+    if (!SSL_get_handshake_rtt(SSL_CONNECTION_GET_SSL(s), &rtt))
```

Comment:
Now this is wrong because you cannot test with `!` for tri-state return. I actually dislike tristate returns for that reason.

Use `if (!TEST_int_gt(SSL_get_handshake_rtt(SSL_CONNECTION_GET_SSL(s), &rtt), 0))`

Reply:
That makes sense, thanks for catching that. I'll edit this to verify that it returns `1`

Reply:
Great, I'll make that change